Basic/Invaders/Assets/Scripts/InvadersGame.cs

```diff
@@ -268,90 +290,109 @@ private void UpdateGameTimer()
 
-            var foundEdge = false;
-            var foundEligibleAlienEnemy = false;
-            UpdateShootingEnemies(ref foundEligibleAlienEnemy, ref foundEdge);
+            UpdateEnemiesResultFlags enemiesResultFlags = UpdateEnemiesResultFlags.None;
+            UpdateShootingEnemies(ref enemiesResultFlags);
 
-            if (!foundEligibleAlienEnemy)
+            if((enemiesResultFlags & UpdateEnemiesResultFlags.ReachedBottom) != 0)
             {
-                CreateAliens();
+                // Force game end as at least one of the enemies have reached the bottom!
+                SetGameEnd(GameOverReason.EnemiesReachedBottom);
+                return;
+            }
+            
+            // If we didn't find any enemies, then spawn some
+            if ((enemiesResultFlags & UpdateEnemiesResultFlags.FoundEnemy) == 0)
+            {
+                CreateEnemies();
                 m_TickPeriodic.Value = 0.2f;
             }
 
-            if (foundEdge)
+            // If the enemies reached the either side of the boundaries, then change the movement direction
+            // And move them to the next row below
+            if ((enemiesResultFlags & UpdateEnemiesResultFlags.ReachedHorizontalBoundary) != 0)
             {
-                m_AlienDirection.Value = -m_AlienDirection.Value;
+                m_EnemyMovingDirection.Value = -m_EnemyMovingDirection.Value;
                 m_TickPeriodic.Value *= 0.95f; // get faster
 
-                var aliensCount = m_Aliens.Count;
-                for (var index = 0; index < aliensCount; index++)
+                var enemiesCount = m_Enemies.Count;
+                for (var index = 0; index < enemiesCount; index++)
                 {
-                    var alien = m_Aliens[index];
-                    alien.transform.Translate(0, k_AlienVerticalMovementOffset, 0);
+                    var enemy = m_Enemies[index];
+                    enemy.transform.Translate(0, k_EnemyVerticalMovementOffset, 0);
                 }
             }
 
             if (m_Saucer == null)
                 if (Random.Range(0, 1.0f) > m_RandomThresholdForSaucerCreation)
-                    CreateSaucer();
+                    CreateSuperEnemy();
         }
     }
 
-    private bool UpdateShootingEnemies(ref bool foundAlien, ref bool foundEdge)
+    private void UpdateShootingEnemies(ref UpdateEnemiesResultFlags flags)
     {
-        var aliensCount = m_Aliens.Count;
-        for (var index = 0; index < aliensCount; index++)
+        flags = UpdateEnemiesResultFlags.None;
+        var enemiesCount = m_Enemies.Count;
+        for (var index = 0; index < enemiesCount; index++)
         {
-            var alien = m_Aliens[index];
-            Assert.IsTrue(alien);
-
-            if (alien.score > 100)
+            var enemy = m_Enemies[index];
+            Assert.IsTrue(enemy != null);
+            if (m_Enemies == null)
+            {
+                continue;
```

Comment:
This should never happen as we assert above. And it is a silent fail? Maybe throw an exception here?